gap_nina.go

```diff
@@ -0,0 +1,178 @@
+//go:build nina || nano_rp2040
+
+package bluetooth
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	ErrConnect = errors.New("could not connect")
+)
+
+// Scan starts a BLE scan.
+func (a *Adapter) Scan(callback func(*Adapter, ScanResult)) error {
+	if a.scanning {
+		return errScanning
+	}
+
+	if err := a.hci.leSetScanEnable(false, true); err != nil {
+		return err
+	}
+
+	if err := a.hci.leSetScanParameters(0x01, 0x0020, 0x0020, 0x00, 0x00); err != nil {
+		return err
+	}
+
+	a.scanning = true
+
+	if err := a.hci.leSetScanEnable(true, false); err != nil {
+		return err
+	}
+
+	for {
+		if err := a.hci.poll(); err != nil {
+			return err
+		}
+
+		switch {
+		case a.hci.advData.reported:
+			if a.hci.advData.typ != 0x04 {
+				a.hci.clearAdvData()
+				continue
+			}
+
+			adf := AdvertisementFields{}
+			if a.hci.advData.eirLength > 64 {
+				if _debug {
+					println("eirLength too long")
+				}
+
+				a.hci.clearAdvData()
+				continue
+			}
+
+			for i := 0; i < int(a.hci.advData.eirLength); {
+				l, t := int(a.hci.advData.eirData[i]), a.hci.advData.eirData[i+1]
+				if l < 1 {
+					break
+				}
+
+				switch t {
+				case 0x02, 0x03:
+					// 16-bit Service Class UUID
+				case 0x06, 0x07:
+					// 128-bit Service Class UUID
+				case 0x08, 0x09:
+					if _debug {
+						println("local name", string(a.hci.advData.eirData[i+2:i+2+l]))
+					}
+
+					adf.LocalName = string(a.hci.advData.eirData[i+2 : i+2+l])
+				case 0xFF:
+					// Manufacturer Specific Data
+				}
+
+				i += l + 1
+			}
+
+			callback(a, ScanResult{
+				Address: Address{MACAddress{MAC: makeAddress(a.hci.advData.peerBdaddr)},
+					a.hci.advData.peerBdaddrType},
+				RSSI: int16(a.hci.advData.rssi),
+				AdvertisementPayload: &advertisementFields{
+					AdvertisementFields: adf,
+				},
+			})
+
+			a.hci.clearAdvData()
+
+		default:
+			if !a.scanning {
+				return nil
+			}
+
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+
+	return nil
+}
+
+func (a *Adapter) StopScan() error {
+	if !a.scanning {
+		return errNotScanning
+	}
+
+	if err := a.hci.leSetScanEnable(false, false); err != nil {
+		return err
+	}
+
+	a.scanning = false
+
+	return nil
+}
+
+// Address contains a Bluetooth MAC address.
+type Address struct {
+	MACAddress
+
+	typ uint8
```

Comment:
What's this `typ`? A device is normally identified just by the address (and possibly the `IsRandom` bit).

Reply:
ah, that should be renamed to random then.

Reply:
Done.